libraries/AC_SafeRTL/AC_SafeRTL.cpp

```diff
@@ -0,0 +1,467 @@
+        return true;
+    }
+
+    // otherwise we'll see how much we could clean up by pruning loops
+    _remove_unacceptable_overlapping_loops();
+
+    int potential_amount_to_prune = 0;
+    for (uint32_t i = 0; i < _prunable_loops.size() - 1; i++) {
+        // add 1 at the end, because a pruned loop is always replaced by one new point.
+        potential_amount_to_prune += _prunable_loops[i].end_index - _prunable_loops[i].start_index + 1;
+    }
+
+    // if pruning could remove 10+ points, prune loops until 10 or more points have been removed (doesn't necessarily prune all loops)
+    if (potential_amount_to_prune >= 10) {
+        _zero_points_by_loops(10);
+        _remove_empty_points();
+        // end by resetting the state of the cleanup methods.
+        _reset_rdp();
+        _reset_pruning();
+        return true;
+    }
+
+    // as a last resort, see if pruning and simplifying together would remove 10+ points.
+    if (potential_amount_to_prune + potential_amount_to_simplify >= 10) {
+        _zero_points_by_simplification_bitmask();
+        _zero_points_by_loops(10);
+        _remove_empty_points();
+        // end by resetting the state of the cleanup methods.
+        _reset_rdp();
+        _reset_pruning();
+        return true;
+    }
+    return false;
+}
+
+/**
+*  Run this method only when preparing to initiate the RTL procedure. Returns a
+*  pointer to the cleaned-up path. Returns nullptr if the cleanup algorithms aren't ready yet.
+*  If this happens, just run this method again a bit later.
+*
+*  Probably best not to run this unless cleanup_ready() is returning true
+*/
+Vector3f* SafeRTL_Path::thorough_cleanup()
+{
+    // apply simplification
+    _zero_points_by_simplification_bitmask();
+
+    // apply pruning
+    _remove_unacceptable_overlapping_loops();
+    _zero_points_by_loops(MAX_PATH_LEN); // prune every single loop
+
+    _remove_empty_points();
+
+    // end by resetting the state of the cleanup methods.
+    _reset_rdp();
+    _reset_pruning();
+
+    return path;
+}
+
+/**
+*   Returns true if the list is empty after popping this point.
+*/
+bool SafeRTL_Path::pop_point(Vector3f& point)
+{
+    if(_last_index == 0){
+        point = path[0];
+        return true;
+    }
+    point =  path[_last_index--];
+    return false;
+}
+
+void SafeRTL_Path::reset_path(Vector3f start)
+{
+    _last_index = 0;
+    path[_last_index] = start;
+    _active = true;
+    _simplification_complete = false;
+    _simplification_clean_until = 0;
+    _pruning_complete = false;
+    _pruning_clean_until = 0;
+}
+
+/**
+*    Simplifies a 3D path, according to the Ramer-Douglas-Peucker algorithm.
+*    Returns the number of items which were removed. end_index is the index of the last element in the path.
+*/
+void SafeRTL_Path::rdp()
+{
+    if (_simplification_complete) {
+        return;
+    }
+    else if (_simplification_stack.is_empty()) {  // if not complete but also nothing to do, we must be restarting.
```

Comment:
formatting guidelines are "} else" instead of putting the else on a new line.  a tiny non-functional thing of course..